File: sdks/php/src/Api/Environment.php
<?php

declare(strict_types=1);

namespace HookSniff\Api;

use HookSniff\Request\HookSniffHttpClient;

class Environment
{
    private HookSniffHttpClient $client;

    public function __construct(HookSniffHttpClient $client)
    {
        $this->client = $client;
    }

    public function list(): array
    {
        return $this->client->request('GET', '/api/v1/environments');
    }

    public function create(array $environmentIn): array
    {
        return $this->client->request('POST', '/api/v1/environments', $environmentIn);
    }

    public function get(string $environmentId): array
    {
        return $this->client->request('GET', "/api/v1/environments/{$environmentId}");
    }

    public function update(string $environmentId, array $environmentPatch): array
    {
        return $this->client->request('PUT', "/api/v1/environments/{$environmentId}", $environmentPatch);
    }

    public function delete(string $environmentId): void
    {
        $this->client->request('DELETE', "/api/v1/environments/{$environmentId}");
    }

    public function listVariables(string $environmentId): array
    {
        return $this->client->request('GET', "/api/v1/environments/{$environmentId}/variables");
    }

    public function getVariable(string $environmentId, string $variableId): array
    {
        return $this->client->request('GET', "/api/v1/environments/{$environmentId}/variables/{$variableId}");
    }

    public function createVariable(string $environmentId, array $variableIn): array
    {
        return $this->client->request('POST', "/api/v1/environments/{$environmentId}/variables", $variableIn);
    }

    public function updateVariable(string $environmentId, string $variableId, array $variableIn): array
    {
        return $this->client->request('PUT', "/api/v1/environments/{$environmentId}/variables/{$variableId}", $variableIn);
    }

    public function deleteVariable(string $environmentId, string $variableId): void
    {
        $this->client->request('DELETE', "/api/v1/environments/{$environmentId}/variables/{$variableId}");
    }

    public function bulkUpsertVariables(string $environmentId, array $bulkIn): array
    {
        return $this->client->request('POST', "/api/v1/environments/{$environmentId}/variables/bulk", $bulkIn);
    }
}

File: sdks/rust/src/api/environment.rs
/*
 * Environment API for HookSniff
 */

use crate::api_internal::HookSniffHttpClient;
use crate::error::Error;
use crate::models::{
    EnvironmentIn, EnvironmentModelOut, EnvironmentPatch, EnvironmentVariableBulkUpsertIn,
    EnvironmentVariableIn, EnvironmentVariableOut,
};

pub struct Environment<'a> {
    client: &'a HookSniffHttpClient,
}

impl<'a> Environment<'a> {
    pub fn new(client: &'a HookSniffHttpClient) -> Self {
        Self { client }
    }

    pub async fn list(&self) -> Result<Vec<EnvironmentModelOut>, Error> {
        self.client.get("/api/v1/environments").await
    }

    pub async fn create(&self, body: &EnvironmentIn) -> Result<EnvironmentModelOut, Error> {
        self.client.post("/api/v1/environments", body).await
    }

    pub async fn get(&self, environment_id: &str) -> Result<EnvironmentModelOut, Error> {
        let path = format!("/api/v1/environments/{}", environment_id);
        self.client.get(&path).await
    }

    pub async fn update(
        &self,
        environment_id: &str,
        body: &EnvironmentPatch,
    ) -> Result<EnvironmentModelOut, Error> {
        let path = format!("/api/v1/environments/{}", environment_id);
        self.client.put(&path, body).await
    }

    pub async fn delete(&self, environment_id: &str) -> Result<(), Error> {
        let path = format!("/api/v1/environments/{}", environment_id);
        self.client.delete(&path).await
    }

    pub async fn list_variables(
        &self,
        environment_id: &str,
    ) -> Result<Vec<EnvironmentVariableOut>, Error> {
        let path = format!("/api/v1/environments/{}/variables", environment_id);
        self.client.get(&path).await
    }

    pub async fn get_variable(
        &self,
        environment_id: &str,
        variable_id: &str,
    ) -> Result<EnvironmentVariableOut, Error> {
        let path = format!(
            "/api/v1/environments/{}/variables/{}",
            environment_id, variable_id
        );
        self.client.get(&path).await
    }

    pub async fn create_variable(
        &self,
        environment_id: &str,
        body: &EnvironmentVariableIn,
    ) -> Result<EnvironmentVariableOut, Error> {
        let path = format!("/api/v1/environments/{}/variables", environment_id);
        self.client.post(&path, body).await
    }

    pub async fn update_variable(
        &self,
        environment_id: &str,
        variable_id: &str,
        body: &EnvironmentVariableIn,
    ) -> Result<EnvironmentVariableOut, Error> {
        let path = format!(
            "/api/v1/environments/{}/variables/{}",
            environment_id, variable_id
        );
        self.client.put(&path, body).await
    }

    pub async fn delete_variable(
        &self,
        environment_id: &str,
        variable_id: &str,
    ) -> Result<(), Error> {
        let path = format!(
            "/api/v1/environments/{}/variables/{}",
            environment_id, variable_id
        );
        self.client.delete(&path).await
    }

    pub async fn bulk_upsert_variables(
        &self,
        environment_id: &str,
        body: &EnvironmentVariableBulkUpsertIn,
    ) -> Result<Vec<EnvironmentVariableOut>, Error> {
        let path = format!("/api/v1/environments/{}/variables/bulk", environment_id);
        self.client.post(&path, body).await
    }
}
